Enable YUV400 JPEG Enc and BGR4 JPEG Enc (api2x)

vcheah opened the pull request:
The commit includes: 
1.  Fixed BGR4 JPEG Enc (api2x)
2. Enable YUV400 JPEG Enc (api1x and api2x)

vcheah commented:
The CD/CI error does not look like it comes from the MR. Anyone has any idea about this?

Sherry-Lin commented:
Yes, there is build regression in recent changes. We are trying to fix it by end of this week.

OttoYang commented on _studio/mfx_lib/shared/src/libmfxsw_encode.cpp:
```diff
@@ -512,6 +512,9 @@ static mfxStatus SetupCache(mfxSession session, const mfxVideoParam& par)
 
     mfxU16 memory_type = mfxU16(par.IOPattern == MFX_IOPATTERN_IN_SYSTEM_MEMORY ? MFX_MEMTYPE_FROM_ENCODE | MFX_MEMTYPE_SYSTEM_MEMORY : MFX_MEMTYPE_FROM_ENCODE | MFX_MEMTYPE_DXVA2_DECODER_TARGET);
 
+    if (par.mfx.CodecId == MFX_CODEC_JPEG && par.mfx.FrameInfo.FourCC == MFX_FOURCC_BGR4)
```
Is this flag required for RGB4?   
And consider moving the logic for codec and format specific restrictions from common file to subclass implementation.

vcheah replied:
The flag is required for BGR4 JPEG Enc case otherwise, the vaattrib are not being set at all (https://github.com/intel/vpl-gpu-rt/blob/79ba6c8ff70aaf15ad9ababc4fdf3e8af611c80b/_studio/shared/src/libmfx_allocator_vaapi.cpp#L204).

 In RGB4 JPEG Enc case, I noticed the condition entered during the WA RGB swapping (https://github.com/intel/vpl-gpu-rt/blob/62ed300338bc6bd7021066bb2900e70217b71155/_studio/mfx_lib/encode_hw/mjpeg/src/mfx_mjpeg_encode_hw.cpp#L1064) before the SetupCache(). After the SetupCache(), MFX_MEMTYPE_VIDEO_MEMORY_ENCODER_TARGET flag flag is never set (similar as BGR4 case).

vcheah replied:
Will also try to check if we can do this within JPEG specific codes.

vcheah replied:
Added a class method and move the logics into JPEG specific codes. 
I have also checked JPEG api2x encoded output for RGB4/BGR4/NV12 are all correct (no corruption).

bai-isaac commented on _studio/mfx_lib/encode_hw/mjpeg/src/mfx_mjpeg_encode_hw.cpp:
```diff
@@ -1195,6 +1202,14 @@ mfxStatus MFXVideoENCODEMJPEG_HW::CheckDevice()
+mfxU16 MFXVideoENCODEMJPEG_HW::UpdateMemType(mfxU16 memory_type)
+{
+#if defined(LINUX)
+    memory_type |= MFX_MEMTYPE_VIDEO_MEMORY_ENCODER_TARGET; // required for libva especially for RGB32
```
could you please teach why we need MFX_MEMTYPE_VIDEO_MEMORY_ENCODER_TARGET for rgb32? Besides, why rgb32, as title is "bgr4"

vcheah replied:
I've simplified the earlier change. I've added the flag to JPEG specific-way and only for BGR4 only since corruption only happens for BGR4 api2x.

OttoYang commented on _studio/mfx_lib/shared/src/libmfxsw_encode.cpp:
```diff
@@ -512,6 +512,8 @@ static mfxStatus SetupCache(mfxSession session, const mfxVideoParam& par)
 
     mfxU16 memory_type = mfxU16(par.IOPattern == MFX_IOPATTERN_IN_SYSTEM_MEMORY ? MFX_MEMTYPE_FROM_ENCODE | MFX_MEMTYPE_SYSTEM_MEMORY : MFX_MEMTYPE_FROM_ENCODE | MFX_MEMTYPE_DXVA2_DECODER_TARGET);
 
+    memory_type = session->m_pENCODE->UpdateMemType(memory_type);
```
It is still confusing to have a variable calculated twice in the same place, and please don't add new redundant codes in other codecs as we don't have such a test plan to cover these changes. Could this be improved in a jpeg-specific way?

vcheah replied:
I've simplified the earlier change. I've added the flag to JPEG specific-way and only for BGR4 only since corruption only happens for BGR4 api2x.

vcheah commented:
Hi @bai-isaac @OttoYang @leyu-yao,
Could you please help to review?

OttoYang reviewed (approved):
LGTM

OttoYang commented:
@vcheah Please help with the pre-check in failure issue resolved in this PR that was detected in the [back merge PR](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723).

vcheah commented:
> @vcheah Please help with the pre-check in failure issue resolved in this PR that was detected in the [back merge PR](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723).

Will take a look.

vcheah commented:
> @vcheah Please help with the pre-check in failure issue resolved in this PR that was detected in the [back merge PR](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723).

Hi @OttoYang 
I have fixed and rebased https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723. I got the pr_verify fixed by updating the commits accordingly with OSPR/Issue/Test/GithubPR information (below).  I notice now only  "Check OS Mergeable" and "Update resources" still the failure.  Do you have any idea about this? 

OSPR: [PUBLIC]333

Issue:  XXXX 
Test: -
Github PR: 333 (Enable YUV400 JPEG Enc and BGR4 JPEG Enc (api2x))

OttoYang commented:
> > @vcheah Please help with the pre-check in failure issue resolved in this PR that was detected in the [back merge PR](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723).
> 
> Hi @OttoYang I have fixed and rebased [intel-innersource/drivers.gpu.media.sdk.lib#8723](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723). I got the pr_verify fixed by updating the commits accordingly with OSPR/Issue/Test/GithubPR information (below). I notice now only "Check OS Mergeable" and "Update resources" still the failure. Do you have any idea about this?
> 
> OSPR: [PUBLIC]333
> 
> Issue: XXXX Test: - Github PR: 333 (Enable YUV400 JPEG Enc and BGR4 JPEG Enc (api2x))

Please fix the issue in this PR, the back merge PR only for verifying.

vcheah commented:
> > > @vcheah Please help with the pre-check in failure issue resolved in this PR that was detected in the [back merge PR](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723).
> > 
> > 
> > Hi @OttoYang I have fixed and rebased [intel-innersource/drivers.gpu.media.sdk.lib#8723](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723). I got the pr_verify fixed by updating the commits accordingly with OSPR/Issue/Test/GithubPR information (below). I notice now only "Check OS Mergeable" and "Update resources" still the failure. Do you have any idea about this?
> > OSPR: [PUBLIC]333
> > Issue: XXXX Test: - Github PR: 333 (Enable YUV400 JPEG Enc and BGR4 JPEG Enc (api2x))
> 
> Please fix the issue in this PR, the back merge PR only for verifying.



> > > @vcheah Please help with the pre-check in failure issue resolved in this PR that was detected in the [back merge PR](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723).
> > 
> > 
> > Hi @OttoYang I have fixed and rebased [intel-innersource/drivers.gpu.media.sdk.lib#8723](https://github.com/intel-innersource/drivers.gpu.media.sdk.lib/pull/8723). I got the pr_verify fixed by updating the commits accordingly with OSPR/Issue/Test/GithubPR information (below). I notice now only "Check OS Mergeable" and "Update resources" still the failure. Do you have any idea about this?
> > OSPR: [PUBLIC]333
> > Issue: XXXX Test: - Github PR: 333 (Enable YUV400 JPEG Enc and BGR4 JPEG Enc (api2x))
> 
> Please fix the issue in this PR, the back merge PR only for verifying.

Hi @OttoYang 
I have updated the commit message accordingly.

vcheah commented:
Hi @OttoYang , 
Could you please help to trigger the innersource build? 
Also, please do let me know if there are anything to do in order to get the code merge.
